jets/e/argon2.c

```diff
@@ -0,0 +1,131 @@
+/* j/5/argon2.c
+**
+*/
+#include "all.h"
+
+#include <argon2.h>
+
+/* functions
+*/
+
+  u3_noun
+  u3qe_argon2(// configuration params,
+              u3_atom out, u3_atom type, u3_atom version,
+              u3_atom threads, u3_atom mem_cost, u3_atom time_cost,
+              u3_atom wik, u3_atom key, u3_atom wix, u3_atom extra,
+              // input params
+              u3_atom wid, u3_atom dat, u3_atom wis, u3_atom sat)
+  {
+    c3_assert(_(u3a_is_cat(out)) && _(u3a_is_cat(type)) &&
+              _(u3a_is_cat(version)) && _(u3a_is_cat(threads)) &&
+              _(u3a_is_cat(mem_cost)) && _(u3a_is_cat(time_cost)) &&
+              _(u3a_is_cat(wik)) && _(u3a_is_cat(wix)) &&
+              _(u3a_is_cat(wid)) && _(u3a_is_cat(wis)));
+
+    // flip endianness for argon2
+    key = u3qc_rev(3, wik, key);
+    extra = u3qc_rev(3, wix, extra);
+    dat = u3qc_rev(3, wid, dat);
+    sat = u3qc_rev(3, wis, sat);
+
+    // atoms to byte arrays
+    c3_y bytes_key[wik];
+    u3r_bytes(0, wik, bytes_key, key);
+    c3_y bytes_extra[wix];
+    u3r_bytes(0, wix, bytes_extra, extra);
+    c3_y bytes_dat[wid];
+    u3r_bytes(0, wid, bytes_dat, dat);
+    c3_y bytes_sat[wis];
+    u3r_bytes(0, wis, bytes_sat, sat);
+
+    c3_y outhash[out];
+    argon2_context context = {
+      outhash,             // output array, at least [digest length] in size
+      out,                 // digest length
+      bytes_dat,           // password array
+      wid,                 // password length
+      bytes_sat,           // salt array
+      wis,                 // salt length
+      bytes_key, wik,      // optional secret data
+      bytes_extra, wix,    // optional associated data
+      time_cost, mem_cost, threads, threads, // performance cost configuration
+      version,             // algorithm version
+      NULL, NULL,          // custom memory allocation / deallocation functions
```

Comment:
We should specify these and, we also need to patch `argon2` to not use `calloc` to allocate it's threading structs.